ports-of-call/robust_utils.hpp

```diff
@@ -64,13 +64,40 @@ PORTABLE_FORCEINLINE_FUNCTION auto ratio(const A &a, const B &b) {
   return a / (b + sgn(b) * SMALL<B>());
 }
 
+template <typename DataType>
+PORTABLE_FUNCTION constexpr DataType reldiff(const DataType actual,
+                                             const DataType expected) {
+  if (actual == expected) {
+    return 0;
+  } else if (expected == 0) {
+    // expected 0, got nonzero, define this to be 100% error
+    return 1;
+  } else {
+    // Note: we do not use std::abs to calculate the absolute value because
+    //       it is not constexpr until C++23, therefore it is not portable
+    //       to GPUs (currently using C++17).
+    const auto signed_reldiff = (actual - expected) / expected;
```

Comment:
Are you married to this particular form? I would prefer something like
```C++
Robust::ratio(2.*(actual - expected), actual + expected);
```
but if you need this exact form, that's fine.

Reply:
The form I used is more appropriate for "I know the right answer and want to compare against it" while your form is more appropriate for "I have two approximate answers and want to see how close they are".  But in practice, they both work well for these kinds of comparisons.  So it's not a huge deal either way.

That being said: see my comment about dropping `reldiff` entirely, because that would supersede this question.